pkg/operator/terminationobserver/termination_observer.go

```diff
@@ -0,0 +1,272 @@
+
+	return c
+}
+
+func (c *TerminationObserver) Register() {
+	registerMetrics(c.getApiServers)
+}
+
+func (c *TerminationObserver) getApiServers() []*apiServerStatus {
+	c.apiServersMutex.RLock()
+	defer c.apiServersMutex.RUnlock()
+	return c.apiServers
+}
+
+// updateApiServerCreationTimestamp updates the creation timestamp for single API server
+func (c *TerminationObserver) updateApiServerCreationTimestamp(name string, currentCreationTimestamp time.Time, isNew bool) {
+	c.apiServersMutex.Lock()
+	defer c.apiServersMutex.Unlock()
+	found := false
+	for i := range c.apiServers {
+		if c.apiServers[i].serverName != name {
+			continue
+		}
+		found = true
+		klog.Infof("Observed API server pod %q termination at %s", c.apiServers[i].serverName, currentCreationTimestamp)
+		if !isNew {
+			c.apiServers[i].terminatedTimestamp = &c.apiServers[i].currentCreatedTimestamp
+		}
+		c.apiServers[i].currentCreatedTimestamp = currentCreationTimestamp
+		break
+	}
+	if !found {
+		klog.Infof("Observed new API server pod %q at %s", name, currentCreationTimestamp)
+		c.apiServers = append(c.apiServers, &apiServerStatus{serverName: name, currentCreatedTimestamp: currentCreationTimestamp})
+	}
+}
+
+func (c *TerminationObserver) apiServerNames() []string {
+	c.apiServersMutex.RLock()
+	defer c.apiServersMutex.RUnlock()
+	var names []string
+	for i := range c.apiServers {
+		names = append(names, c.apiServers[i].serverName)
+	}
+	return names
+}
+
+func (c *TerminationObserver) addApiServerTerminationEvent(event *corev1.Event) {
+	c.apiServersMutex.Lock()
+	defer c.apiServersMutex.Unlock()
+	for i := range c.apiServers {
+		if c.apiServers[i].serverName != event.InvolvedObject.Name {
+			continue
+		}
+		c.apiServers[i].terminationEvents = append(c.apiServers[i].terminationEvents, event)
+		sort.Slice(c.apiServers[i].terminationEvents, func(j, k int) bool {
+			return c.apiServers[i].terminationEvents[j].LastTimestamp.Before(&c.apiServers[i].terminationEvents[k].LastTimestamp)
+		})
+		klog.Infof("Observed termination event %q for API server %q: %s", event.Reason, c.apiServers[i].serverName, event.Message)
+		break
+	}
+}
+
+func (c *TerminationObserver) getApiServerCreationTimestamp(name string) *time.Time {
+	c.apiServersMutex.RLock()
+	defer c.apiServersMutex.RUnlock()
+	for i := range c.apiServers {
+		if c.apiServers[i].serverName != name {
+			continue
+		}
+		return &c.apiServers[i].currentCreatedTimestamp
+	}
+	return nil
+}
+
+func (c *TerminationObserver) sync() error {
+	podList, err := c.podsGetter.Pods(c.targetNamespace).List(metav1.ListOptions{LabelSelector: "app=openshift-kube-apiserver"})
+	if err != nil {
+		return fmt.Errorf("unable to list pods in %q namespace: %v", c.targetNamespace, err)
+	}
+
+	for _, pod := range podList.Items {
+		// get the last seen creation timestamp for this api server
+		lastSeenCreationTimestamp := c.getApiServerCreationTimestamp(pod.Name)
```

Comment:
s/getApiServerCreationTimestamp/lastCreationTimestamp/

Reply:
no, this is "current" creation timestamp